ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -0,0 +1,210 @@
+                stream_iter(iterator.enumerate().map(|(n, x)| {
+                    let mut fc = f.clone();
+                    async move {
+                        // compare current element against pivot
+                        let sh_comp = compare_gt(
+                            pctx.narrow(&Step::Compare)
+                                .set_total_records(b_r - (b_l + 1)),
+                            RecordId::from(n),
+                            // apply key extraction function f to element x
+                            fc(x),
+                            pivot,
+                        )
+                        .await?;
+
+                        // reveal outcome of comparison
+                        sh_comp
+                            .reveal(
+                                pctx.narrow(&Step::Reveal)
+                                    .set_total_records(b_r - (b_l + 1)),
+                                RecordId::from(n),
+                            )
+                            .await
+                    }
+                })),
+            )
+            .try_collect::<Vec<_>>()
+            .await?;
+
+            // swap elements based on comparisons
+            // i is index of first element larger than pivot
+            let mut i = b_l + 1;
+            for j in b_l + 1..b_r {
+                // desc = true will flip the order
+                if comp[j - (b_l + 1)] == Boolean::from(false ^ desc) {
+                    list.swap(i, j);
+                    i += 1;
+                }
+            }
+
+            // put pivot to index i-1
+            list.swap(i - 1, b_l);
+
+            // push recursively calls to quicksort function on stack
+            if i > b_l + 1 {
+                stack.push((ctx.narrow(&Step::Left), b_l, i - 1));
+            }
+            if i + 1 < b_r {
+                stack.push((ctx.narrow(&Step::Right), i, b_r));
+            }
+        }
+    }
+
+    // no error happened, sorted successfully
+    Ok(())
+}
+
+#[cfg(all(test, unit_test))]
+pub mod tests {
+    use rand::Rng;
+
+    use crate::{
+        error::Error,
+        ff::{boolean_array::BA64, Field},
+        protocol::{context::Context, ipa_prf::quicksort::quicksort_by_key_insecure},
+        rand::thread_rng,
+        secret_sharing::replicated::semi_honest::AdditiveShare,
+        test_executor::run,
+        test_fixture::{Reconstruct, Runner, TestWorld},
+    };
+
+    pub async fn quicksort_insecure_test<C>(
+        ctx: C,
+        list: &[AdditiveShare<BA64>],
+        desc: bool,
+    ) -> Result<Vec<AdditiveShare<BA64>>, Error>
+    where
+        C: Context,
+    {
+        let mut list_mut = list.to_vec();
+        quicksort_by_key_insecure(ctx, &mut list_mut[..], desc, |x| x).await?;
+        let mut result: Vec<AdditiveShare<BA64>> = vec![];
+        list_mut.iter().for_each(|x| result.push(x.clone()));
+        Ok(result)
+    }
+
+    #[test]
+    fn test_quicksort_insecure_semi_honest() {
```

Comment:
do we also want to test on edge cases (all equal, sorted in opposite direction, empty list) etc?

Reply:
added all these test cases